python_modules/dagster/dagster/_core/definitions/schedule_definition.py

```diff
@@ -198,28 +194,35 @@ def __init__(
         self._repository_name = repository_name
         self._schedule_name = schedule_name
 
-        self._resources_cm = build_resources(resources or {})
-
-        self._resources = self._resources_cm.__enter__()
-
-        self._resources_contain_cm = isinstance(self._resources, IContainsGenerator)
+        # Wait to set resources unless they're accessed
+        self._resource_defs = resources
+        self._resources = None
+        self._resources_contain_cm = None
         self._cm_scope_entered = False
 
     def __enter__(self) -> "ScheduleEvaluationContext":
         self._cm_scope_entered = True
         return self
 
     def __exit__(self, *exc) -> None:
-        self._resources_cm.__exit__(*exc)  # pylint: disable=no-member
         self._exit_stack.close()
         self._logger = None
 
-    def __del__(self) -> None:
-        if self._resources_contain_cm and not self._cm_scope_entered:
-            self._resources_cm.__exit__(None, None, None)  # pylint: disable=no-member
-
     @property
     def resources(self) -> Resources:
+        if not self._resources:
+            from dagster._core.definitions.scoped_resources_builder import (
+                IContainsGenerator,
+            )
+            from dagster._core.execution.build_resources import build_resources
+
+            instance = self.instance if self._instance or self._instance_ref else None
+
+            resources_cm = build_resources(resources=self._resource_defs or {}, instance=instance)
+            self._resources = self._exit_stack.enter_context(resources_cm)
+
+            self._resources_contain_cm = isinstance(self._resources, IContainsGenerator)
```

Comment:
does this still need to be state / do we use this anywhere outside this function now?

Reply:
We use it in this function, but in cases where the `if` is not triggered, e.g. on subsequent calls to `.resources`

Reply:
ah wait, yeah, can totally move it to a local